Upload from StringIO object

donoghuc opened the pull request:
This commit adds the ability to upload a StringIO object as a source "file". This allows data from memory to be uploaded to a remote file location via winrm.

donoghuc commented:
The ability to copy from an in-memory buffer is supported in [net-scp](https://github.com/net-ssh/net-scp#synopsis). Adding it to winrm-fs would be nice feature parity. Happy to add to the documentation and spec tests if idea is approved. Thanks.

MikaelSmith commented:
The BOLT ticket number probably isn't very helpful in the message.

lucywyman commented:
@mwrock @sneal  Seem to be the maintainers?  We'd love to get your glances and feedback sometime in the next week or 2! Though can always extend the gem ourselves for our project ([Bolt](https://github.com/puppetlabs/bolt)) if need be!

mwrock commented on lib/winrm-fs/core/file_transporter.rb:
```diff
@@ -75,9 +75,8 @@ def upload(locals, remote)
-
           elapsed1 = Benchmark.measure do
-            files = make_files_hash(Array(locals), remote)
+            files = make_files_hash(locals.is_a?(StringIO) ? [locals] : Array(locals), remote)
```
how about:
```
files = make_files_hash([locals].flatten, remote)
```

mwrock commented on lib/winrm-fs/core/file_transporter.rb:
```diff
@@ -188,11 +187,10 @@ def add_directory_hash!(hash, dir, remote)
-
           hash[sha1sum(local)] = {
-            'src'   => local,
+            'src'   => local.is_a?(StringIO) ? { src: local, filename: remote } : local,
```
could you keep this line as is and check for `StringIO` instead of `Hash`?

mwrock commented on lib/winrm-fs/core/file_transporter.rb:
```diff
@@ -469,8 +474,12 @@ def stream_upload_file(src, dest, &block)
           elapsed = Benchmark.measure do
-            File.open(src, 'rb') do |io|
-              chunks, bytes = stream_upload(io, dest, &block)
+            if src.is_a?(Hash)
```
if `src` remained a `StringIO` as suggested above, I think you could leave this addition out.

donoghuc replied:
Unfortunately `File.open` will error out if `src` is a `StringIO` object.

donoghuc commented:
@mwrock Thank you for the review. I updated the code based on your comments and added a note in the README. I realized that with the current implementation, only a single `StringIO` object should be allowed to be uploaded at a time because the destination will be a single path on the remote target. I added some logic to enforce that and added a few more integration tests to illustrate expected behavior. For now uploading a single object at a time suites our needs.

MikaelSmith commented on lib/winrm-fs/core/file_transporter.rb:
```diff
@@ -67,17 +75,16 @@ def initialize(shell, opts = {})
-        # @param locals [Array<String>,String] one or more local file or
-        #   directory paths
+        # @param locals [Array<String>,String,StringIO] one or more
+        #   local file or directory paths, StringIO objects also accepted
```
Probably should clarify only a single StringIO object is valid.

mwrock replied:
I think we are good here since it explicitly states `StringIO` and not `Array<StringIO>`

mwrock commented:
This all looks good! I'm gonna do a bit of curation here and get a minor release out tonight or at least before the end of the 3 day US Holiday weekend. Thanks again for this great contribution!!

mwrock commented:
released 1.3.0!
